contrib_bots/bots_test_lib.py

```diff
@@ -44,19 +44,41 @@ def mock_test(self, messages, message_handler, bot_response):
             for (message, response) in zip(messages, bot_response):
                 # Send message to the concerned bot
                 message_handler.handle_message(message, MockClass(), StateHandler())
-
-                # Check if BotHandlerApi is sending a reply message.
-                # This can later be modified to assert the contents of BotHandlerApi.send_message
+                # Check if the bot is sending a reply message.
+                # 'response' is a string here.
                 instance.send_reply.assert_called_with(message, response)
 
+    def mock_test_send_message(self, messages, message_handler, bot_response):
+        # message_handler is of type Any, since it can contain any bot's
+        # handler class. Eventually, we want bot's handler classes to
+        # inherit from a common prototype specifying the handle_message
+        # function.
+        # type: (List[Dict[str, Any]], Any, Dict[str, str]) -> None
+        # Mocking BotHandlerApi
+        with patch('contrib_bots.bot_lib.BotHandlerApi') as MockClass:
+            instance = MockClass.return_value
+
+            for (message, response) in zip(messages, bot_response):
+                # Send message to the concerned bot
+                message_handler.handle_message(message, MockClass(), StateHandler())
+                # Check if the bot is sending a message via `send_message` function.
+                # Where response is a dictionary here.
+                instance.send_message.assert_called_with(response)
+
     def bot_to_run(self, bot_module):
         # Returning Any, same argument as in mock_test function.
         # type: (str) -> Any
         lib_module = get_lib_module(bot_module)
         message_handler = lib_module.handler_class()
         return message_handler
 
-    def bot_test(self, messages, bot_module, bot_response):
+    def bot_test(self, messages, bot_module, bot_response, test_function):
```

Comment:
Regarding code structuring, how do you feel about handling `bot_test` as some sort of initializer that simply does stuff like setting `message_handler = self.bot_to_run(bot_module)`, since that only needs to be done once. We can then merge the function with `assert_bot_output`. Once that's done, functions like `mock_test_send_message` and `mock_test_send_reply` could be called on their own in the bot's tests, so we wouldn't need the `test_function` parameter anymore. This would require a change of the existing tests, but I think it's a cleaner and more understandable approach.